### ISSUE TITLE
Fatal exception when creating a new release

### ISSUE THREAD

**Issue by peters** [Squirrel/OldSquirrelForWindows#113]
Version: 0.6.17-beta

Shimmer stacktrace:

``` cs
Unhandled Exception: 
System.ArgumentException: An item with the same key has already been added.
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector)
   at Shimmer.Core.ReleasePackage.addDeltaFilesToContentTypes(String rootDirectory) in c:\Users\peters\Documents\GitHub2\Shimmer\src\Shimmer.Core\ReleasePackage.cs:line 274
   at Shimmer.Core.ReleasePackage.CreateReleasePackage(String outputFile, String packagesRootDir, Func`2 releaseNotesProcessor) in 
c:\Users\peters\Documents\GitHub2\Shimmer\src\Shimmer.Core\ReleasePackage.cs:line 115
   at CreateReleasePackage.Program.Main(String[] args) in c:\Users\peters\Documents\GitHub2\Shimmer\src\CreateReleasePackage\Program.cs:line 30
```

Shimmer release build output:

``` ps

Shimmer: Using packages directory 

CreateReleasePackage.exe : 
At Shimmer.0.6.17-beta\tools\commands.psm1:69 char:26
+         $releaseOutput = & $createReleasePackageExe -o $releaseDir -p $packageDi ...
+                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError

Unhandled Exception: 
System.ArgumentException: An item with the same key has already been added.
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector)
   at Shimmer.Core.ReleasePackage.addDeltaFilesToContentTypes(String rootDirectory) in c:\Users\peters\Documents\GitHub2\Shimmer\src\Shimmer.Core\ReleasePackage.cs:line 274
   at Shimmer.Core.ReleasePackage.CreateReleasePackage(String outputFile, String packagesRootDir, Func`2 releaseNotesProcessor) in 
c:\Users\peters\Documents\GitHub2\Shimmer\src\Shimmer.Core\ReleasePackage.cs:line 115
   at CreateReleasePackage.Program.Main(String[] args) in c:\Users\peters\Documents\GitHub2\Shimmer\src\CreateReleasePackage\Program.cs:line 30


You cannot call a method on a null-valued expression.
At Shimmer.0.6.17-beta\tools\commands.psm1:71 char:9
+         $packages = $releaseOutput.Split(";")
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull

Cannot index into a null array.
At Shimmer.0.6.17-beta\tools\commands.psm1:72 char:9
+         $fullRelease = $packages[0].Trim()
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : NullArray


Shimmer: Full release: 

Shimmer: Creating installer for 
Shimmer: Running candle.exe
Windows Installer Xml Compiler version 3.6.3303.0
Copyright (C) Outercurve Foundation. All rights reserved.

template.wxs

template.wxs(34) : error CNDL0006 : The Payload/@SourceFile attribute's value cannot be an empty string.  
If you want the value to be null or empty, simply remove the entire attribute.
Shimmer: Running light.exe
light.exe : error LGHT0103 : The system cannot find the file '\\template.wixobj' with type 'Source'.
PM> 
```

**Comment by peters** [comment]
Seems related to https://github.com/github/Shimmer/pull/100

**Comment by shiftkey** [comment]
@peters can you share a sample .nupkg that is triggering this behaviour?

**Comment by shiftkey** [comment]
Specific chunk of code:

``` csharp
var doc = new XmlDocument();
var path = Path.Combine(rootDirectory, "[Content_Types].xml");
doc.Load(path);

var typesElement = doc.FirstChild.NextSibling;
if (typesElement.Name.ToLowerInvariant() != "types") {
    throw new Exception("Invalid ContentTypes file, expected root node should be 'Types'");
}

var existingTypes = typesElement.ChildNodes.OfType<XmlElement>()
            .ToDictionary(k => k.GetAttribute("Extension").ToLowerInvariant(), 
                          k => k.GetAttribute("ContentType").ToLowerInvariant());

```

Where that `[Content_Types].xml` file looks something like this:

``` xml
<?xml version="1.0" encoding="utf-8"?>
<Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types">
  <Default Extension="rels" ContentType="application/vnd.openxmlformats-package.relationships+xml" />
  <Default Extension="nuspec" ContentType="application/octet" />
  <Default Extension="exe" ContentType="application/octet" />
  <Default Extension="psmdcp" ContentType="application/vnd.openxmlformats-package.core-properties+xml" />
</Types>
```

I'm not sure what's changed in here, so getting my hands on that data would be an excellent help to isolating this issue.

**Comment by peters** [comment]
[Content_Types].xml:

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types">
<Default ContentType="application/vnd.openxmlformats-package.relationships+xml" Extension="rels"/>
  <Default ContentType="application/octet" Extension="nuspec"/>
  <Default ContentType="application/octet" Extension="html"/>
  <Default ContentType="application/octet" Extension="nsi"/>
  <Default ContentType="application/octet" Extension="css"/>
  <Default ContentType="application/octet" Extension="otf"/>
  <Default ContentType="application/octet" Extension="png"/>
  <Default ContentType="application/octet" Extension="js"/>
  <Default ContentType="application/octet" Extension="json"/>
  <Default ContentType="application/octet" Extension="md"/>
  <Default ContentType="application/octet" Extension="yml"/>
  <Default ContentType="application/octet" Extension="eot"/>
  <Default ContentType="application/octet" Extension="svg"/>
  <Default ContentType="application/octet" Extension="ttf"/>
  <Default ContentType="application/octet" Extension="woff"/>
  <Default ContentType="application/octet" Extension="jpg"/>
  <Default ContentType="application/octet" Extension="less"/>
  <Default ContentType="application/octet" Extension="coffee"/>
  <Default ContentType="application/octet" Extension="map"/>
  <Default ContentType="application/octet" Extension="txt"/>
  <Default ContentType="application/octet" Extension="gif"/>
  <Default ContentType="application/octet" Extension="markdown"/>
  <Default ContentType="application/octet" Extension="jade"/>
  <Default ContentType="application/octet" Extension="sh"/>
  <Default ContentType="application/octet" Extension="text"/>
  <Default ContentType="application/octet" Extension="pl"/>
  <Default ContentType="application/octet" Extension="dphpd"/>
  <Default ContentType="application/octet" Extension="php"/>
  <Default ContentType="application/octet" Extension="refine"/>
  <Default ContentType="application/octet" Extension="wav"/>
  <Default ContentType="application/octet" Extension="mdown"/>
  <Default ContentType="application/octet" Extension="bar"/>
  <Default ContentType="application/octet" Extension="template"/>
  <Default ContentType="application/octet" Extension="ico"/>
  <Default ContentType="application/octet" Extension="dll"/>
  <Default ContentType="application/octet" Extension="pak"/>
  <Default ContentType="application/octet" Extension="exe"/>
  <Default ContentType="application/vnd.openxmlformats-package.core-properties+xml" Extension="psmdcp"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/bootstrap/CNAME"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/bootstrap/LICENSE"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/hammerjs/LICENSE"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/iscroll/LICENSE"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/jqBootstrapValidation/LICENSE"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/jquery-ui/MANIFEST"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/page/Makefile"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/requirejs/LICENSE"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/requirejs/tests/bar"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/requirejs/tests/foo"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/requirejs/tests/doh/LICENSE"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/requirejs/tests/doh/README"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/requirejs/tests/doh/_sounds/LICENSE"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/requirejs/tests/toUrl/sub/noext"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/requirejs-domready/LICENSE"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/requirejs-plugins/examples/js/foo"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/requirejs-text/LICENSE"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/select2/LICENSE"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/underscore/CNAME"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/underscore/LICENSE"/>
  <Override ContentType="application/octet" PartName="/Content/frontend/js/library/underscore/Rakefile"/>
</Types>
```

**Comment by peters** [comment]
@shiftkey Sure, which email account should i send it to?

**Comment by shiftkey** [comment]
@peters the one in [my profile](https://github.com/shiftkey)

And wow, that's an interesting xml file - what sort of project is it?

**Comment by peters** [comment]
@shiftkey You got mail! It's a small app handling payment processing using cefsharp. The reason why there's a lot of different content types is because it's a debug build where all the bower .git repositories are not compiled via grunt/r.js ;)

**Comment by shiftkey** [comment]
@peters I suspect it's related to the `Override` elements not having an `Extension` element and thus reverting to empty string.

Evidently we need to make this support both `Default` and `Override` elements - shouldn't be too hard to implement. Thanks for sharing.

**Comment by peters** [comment]
:+1:

**Comment by peters** [comment]
@shiftkey Content directory is not extracted to the application install path. See the nuspec i sent you per email. For the application to work, it requires everything in Content folder to be present.

**Comment by shiftkey** [comment]
@peters I'm not sure I follow. Is something not working with how we are porting content across?

I haven't had a chance to run that `.nupkg` through the process - let me know what I should be looking for.

**Comment by peters** [comment]
If you take a look inside the nupkg i sent you, you can see the Content directory. That directory is not extracted during installation (it's not present in the final path after a successfull installation). My program depends on all those files. If you open the nuspec, you can see those dependencies derive from `<files />`

**Comment by shiftkey** [comment]
@peters As that's a different thing to the initial issue discussed, I've moved it over to #125 

Please discuss that issue over there. I'll be marking this specific issue as resolved when the relevant pull request is merged.

**Comment by shiftkey** [comment]
Fixed in `0.6.20`
